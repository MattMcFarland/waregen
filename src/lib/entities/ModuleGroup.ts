import idx from "idx";
import { X4Entity, X4EntityType } from "./X4Entity";
import {
  GroupEntity as ModuleGroupEntity,
  SelectEntity,
  GroupEntity
} from "../XMLTypes/X4LibraryModuleGroups";

export type ModuleGroupEntity = ModuleGroupEntity;
export type ModuleGroupSelectEntity = SelectEntity;

export interface ModuleGroupConstructOptions {
  name: string;
  macros?: string[];
}

export class ModuleGroup extends X4Entity<ModuleGroupEntity> {
  private proxy: any;
  private _macros: string[] = [];
  get name() {
    return this.xmlDef.Attributes.name;
  }
  set name(v: string) {
    if (undefined == v) throw new TypeError("name must be set");
    if (typeof v !== "string")
      throw new TypeError(`${v}:${typeof name} must be a string`);
    this.xmlDef.Attributes.name = v;
  }
  get macros() {
    return this.proxy;
  }
  set macros(v: string[]) {
    const xmlDef = this.xmlDef;
    const handler = {
      set(target: string[], key: string, value: string) {
        xmlDef.select = [
          ...((xmlDef.select && xmlDef.select) || []),
          mapName(value)
        ];
        return Reflect.set(target, key, value);
      },
      deleteProperty(target: string[], key: string) {
        return Reflect.deleteProperty(target, key);
      }
    };
    this.proxy = new Proxy(this._macros, handler);
    Object.assign(this.proxy, v);
  }
  constructor(
    constructOptions?: ModuleGroupConstructOptions,
    xmlDefOptions?: ModuleGroupEntity
  ) {
    super(
      X4EntityType.LIBRARY_MODULE_GROUPS,
      "group",
      xmlDefOptions || {
        Attributes: {
          name: ""
        },
        select: []
      }
    );
    if (constructOptions) {
      this.name = constructOptions.name || "";
      this.macros = constructOptions.macros || [];
    }
  }
}

export function mapName(name: string): any {
  return {
    Attributes: {
      macro: name
    }
  };
}

export function findInXmlDef(xmlDef: GroupEntity, value: string) {
  const macrosArray = <SelectEntity[]>idx(xmlDef, _ => _.select);
  return macrosArray.find((e: SelectEntity) => e.Attributes.macro === value);
}

export default ModuleGroup;
