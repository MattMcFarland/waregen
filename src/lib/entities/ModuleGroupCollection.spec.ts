import { ModuleGroupCollection } from "./ModuleGroupCollection";
import { ModuleGroup, ModuleGroupConstructOptions } from "./ModuleGroup";

const groupTest = (
  name: string,
  macros: string[]
): ModuleGroupConstructOptions => ({
  name,
  macros
});

describe("ModuleGroupCollection", () => {
  describe("constructor", () => {
    const testCollection = new ModuleGroupCollection([
      new ModuleGroup(groupTest("group_1", ["foo", "bar"])),
      new ModuleGroup(groupTest("group_2", ["foo", "bar"])),
      new ModuleGroup(groupTest("group_3", ["foo", "bar"]))
    ]);
    test("success", () => {
      expect(testCollection.extract("group_1")).toBeDefined();
      expect(testCollection.extract("group_2")).toBeDefined();
      expect(testCollection.extract("group_3")).toBeDefined();
    });
    test("instance.toXml()", () => {
      expect(testCollection.toXml()).toMatchInlineSnapshot(`
"<groups>
  <group name=\\"group_1\\">
    <select macro=\\"foo\\"/>
    <select macro=\\"bar\\"/>
  </group>
  <group name=\\"group_2\\">
    <select macro=\\"foo\\"/>
    <select macro=\\"bar\\"/>
  </group>
  <group name=\\"group_3\\">
    <select macro=\\"foo\\"/>
    <select macro=\\"bar\\"/>
  </group>
</groups>"
`);
    });
    test("instance.toJson()", () => {
      expect(testCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"group\\":[{\\"Attributes\\":{\\"name\\":\\"group_1\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"foo\\"}},{\\"Attributes\\":{\\"macro\\":\\"bar\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_2\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"foo\\"}},{\\"Attributes\\":{\\"macro\\":\\"bar\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_3\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"foo\\"}},{\\"Attributes\\":{\\"macro\\":\\"bar\\"}}]}]}"`
      );
    });
  });
  describe("instance.add()", () => {
    const testCollection = new ModuleGroupCollection();
    testCollection.add(new ModuleGroup(groupTest("group_1", ["foo", "bar"])));
    test("instance.toXml()", () => {
      expect(testCollection.toXml()).toMatchInlineSnapshot(`
"<groups>
  <group name=\\"group_1\\">
    <select macro=\\"foo\\"/>
    <select macro=\\"bar\\"/>
  </group>
</groups>"
`);
    });
    test("instance.toJson()", () => {
      expect(testCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"group\\":[{\\"Attributes\\":{\\"name\\":\\"group_1\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"foo\\"}},{\\"Attributes\\":{\\"macro\\":\\"bar\\"}}]}]}"`
      );
    });
  });
  describe("instance.remove()", () => {
    const testCollection = new ModuleGroupCollection();
    testCollection.add(new ModuleGroup(groupTest("group_1", ["foo", "bar"])));
    testCollection.add(new ModuleGroup(groupTest("group_2", ["foo", "bar"])));
    testCollection.remove("group_1");
    test("instance.toXml()", () => {
      expect(testCollection.toXml()).toMatchInlineSnapshot(`
"<groups>
  <group name=\\"group_2\\">
    <select macro=\\"foo\\"/>
    <select macro=\\"bar\\"/>
  </group>
</groups>"
`);
    });
    test("instance.toJson()", () => {
      expect(testCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"group\\":[{\\"Attributes\\":{\\"name\\":\\"group_2\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"foo\\"}},{\\"Attributes\\":{\\"macro\\":\\"bar\\"}}]}]}"`
      );
    });
  });
  describe("instance.replace()", () => {
    const testCollection = new ModuleGroupCollection();
    testCollection.add(new ModuleGroup(groupTest("group_1", ["foo", "bar"])));
    testCollection.add(new ModuleGroup(groupTest("group_2", ["foo", "bar"])));
    const replacement = new ModuleGroup(groupTest("group_2", ["replaced!!"]));
    testCollection.replace(replacement);
    test("instance.toXml()", () => {
      expect(testCollection.toXml()).toMatchInlineSnapshot(`
"<groups>
  <group name=\\"group_1\\">
    <select macro=\\"foo\\"/>
    <select macro=\\"bar\\"/>
  </group>
  <group name=\\"group_2\\">
    <select macro=\\"replaced!!\\"/>
  </group>
</groups>"
`);
    });
    test("instance.toJson()", () => {
      expect(testCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"group\\":[{\\"Attributes\\":{\\"name\\":\\"group_1\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"foo\\"}},{\\"Attributes\\":{\\"macro\\":\\"bar\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_2\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"replaced!!\\"}}]}]}"`
      );
    });
  });
  describe("instance.import()", () => {
    const testXml = () => {
      return `
      <groups>
        <group name="group_1">
          <select macro="group_1_macro" />
        </group>
        <group name="group_2">
          <select macro="group_2_macro" />
          <select macro="group_2_macro2" />
          <select macro="group_2_macro3" />
        </group>
        <group name="group_3">
          <select macro="group_3_macro" />
          <select macro="group_3_macro2" />
          <select macro="group_3_macro3" />
        </group>
        <group name="group_4">
        <select macro="group_4_macro" />
        <select macro="group_4_macro2" />
        <select macro="group_4_macro3" />
        </group>
        <group name="group_5">
          <select macro="group_5_macro" />
          <select macro="group_5_macro2" />
          <select macro="group_5_macro3" />
        </group>
      </groups>
      `;
    };
    test("toXml()", async () => {
      const testCollection = new ModuleGroupCollection();
      await testCollection.import(testXml());
      expect(testCollection.toXml()).toMatchSnapshot();
    });
    test("toJson()", async () => {
      const testCollection = new ModuleGroupCollection();
      await testCollection.import(testXml());
      expect(testCollection.toJson()).toMatchSnapshot();
    });
  });
});
