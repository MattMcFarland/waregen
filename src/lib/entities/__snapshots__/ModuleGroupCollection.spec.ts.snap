// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleGroupsCollection instance.import() toJson() 1`] = `"{\\"group\\":[{\\"Attributes\\":{\\"name\\":\\"group_1\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"group_1_macro\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_2\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"group_2_macro\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_2_macro2\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_2_macro3\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_3\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"group_3_macro\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_3_macro2\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_3_macro3\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_4\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"group_4_macro\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_4_macro2\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_4_macro3\\"}}]},{\\"Attributes\\":{\\"name\\":\\"group_5\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"group_5_macro\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_5_macro2\\"}},{\\"Attributes\\":{\\"macro\\":\\"group_5_macro3\\"}}]}]}"`;

exports[`ModuleGroupsCollection instance.import() toXml() 1`] = `
"<groups>
  <group name=\\"group_1\\">
    <select macro=\\"group_1_macro\\"/>
  </group>
  <group name=\\"group_2\\">
    <select macro=\\"group_2_macro\\"/>
    <select macro=\\"group_2_macro2\\"/>
    <select macro=\\"group_2_macro3\\"/>
  </group>
  <group name=\\"group_3\\">
    <select macro=\\"group_3_macro\\"/>
    <select macro=\\"group_3_macro2\\"/>
    <select macro=\\"group_3_macro3\\"/>
  </group>
  <group name=\\"group_4\\">
    <select macro=\\"group_4_macro\\"/>
    <select macro=\\"group_4_macro2\\"/>
    <select macro=\\"group_4_macro3\\"/>
  </group>
  <group name=\\"group_5\\">
    <select macro=\\"group_5_macro\\"/>
    <select macro=\\"group_5_macro2\\"/>
    <select macro=\\"group_5_macro3\\"/>
  </group>
</groups>"
`;
